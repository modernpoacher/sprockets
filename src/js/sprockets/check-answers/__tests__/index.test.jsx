import React from 'react'
import PropTypes from 'prop-types'
import {
  snapshotOf
} from 'react-component-snapshot'
import classnames from 'classnames'

import '@testing-library/jest-dom'

import {
  render
} from '@testing-library/react'

import {
  getInstanceFromContainerElement
} from 'react-component-instance'

import Super from '#sprockets/sprockets'
import Sprocket from '#sprockets/sprockets/check-answers'

import {
  MOCK_CHECK_ANSWERS
} from './definitions.mjs'

jest.mock('classnames', () => jest.fn().mockReturnValue('MOCK CLASSNAME'))

/**
 *  @param {{
 *    to: string | { pathname: string },
 *    children: React.ReactNode | React.ReactNode[]
 *  }} prop
 *  @returns {React.JSX.Element}
 */
function MockLink ({ to, children }) {
  if (typeof to === 'string') {
    return (
      <a href={to} className='mock-link'>
        {children}
      </a>
    )
  }

  const {
    pathname
  } = to

  return (
    <a href={pathname} className='mock-link'>
      {children}
    </a>
  )
}

MockLink.propTypes = {
  to: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      pathname: PropTypes.string
    })
  ]),
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(
      PropTypes.node
    )
  ])
}

jest.mock('react-router', () => {
  return {
    __esModule: true,
    Link: MockLink
  }
})

describe('#sprockets/sprockets/check-answers', () => {
  describe('<Sprocket />', () => {
    describe('With required props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: sprocket
          }
        } = render(
          <Sprocket />
        )

        expect(sprocket)
          .toBeNull()
      })

      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: sprocket
          }
        } = render(
          <Sprocket />
        )

        expect(snapshotOf(sprocket))
          .toMatchSnapshot()
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: sprocket
          }
        } = render(
          <Sprocket
            title='MOCK TITLE'
            checkAnswers={MOCK_CHECK_ANSWERS}
          />
        )

        expect(sprocket)
          .toBeInstanceOf(HTMLDivElement)
      })

      describe('Always', () => {
        /**
         *  @type {undefined | jest.SpyInstance}
         */
        let getClassNameSpy

        /**
         *  @type {undefined | jest.SpyInstance}
         */
        let renderTitleSpy

        /**
         *  @type {undefined | jest.SpyInstance}
         */
        let renderGroupSpy

        beforeEach(() => {
          getClassNameSpy = jest.spyOn(Sprocket.prototype, 'getClassName')

          renderTitleSpy = jest.spyOn(Sprocket.prototype, 'renderTitle')

          renderGroupSpy = jest.spyOn(Sprocket.prototype, 'renderGroup')

          render(
            <Sprocket
              title='MOCK TITLE'
              checkAnswers={MOCK_CHECK_ANSWERS}
            />
          )
        })

        it('invokes `getClassName`', () => {
          expect(getClassNameSpy)
            .toHaveBeenCalled()
        })

        it('invokes `renderTitle`', () => {
          expect(renderTitleSpy)
            .toHaveBeenCalled()
        })

        it('invokes `renderGroup`', () => {
          expect(renderGroupSpy)
            .toHaveBeenCalled()
        })
      })

      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: sprocket
          }
        } = render(
          <Sprocket
            title='MOCK TITLE'
            checkAnswers={MOCK_CHECK_ANSWERS}
          />
        )

        expect(snapshotOf(sprocket))
          .toMatchSnapshot()
      })
    })

    describe('`getClassName()`', () => {
      it('invokes `classnames`', () => {
        /**
         *  Ensure `super.getClassName()` returns a value
         */
        const getClassNameSpy = jest.spyOn(Super.prototype, 'getClassName').mockReturnValue('MOCK CLASSNAME')

        const {
          container
        } = render(
          <Sprocket />
        )

        const instance = getInstanceFromContainerElement(container)

        /**
         *  Ensure it is reset after render
         */
        classnames.mockClear()

        /**
         *  Ensure it is reset after render
         */
        getClassNameSpy.mockClear()

        instance.getClassName()

        expect(classnames)
          .toHaveBeenCalledWith('MOCK CLASSNAME', 'check-answers')
      })
    })
  })
})
