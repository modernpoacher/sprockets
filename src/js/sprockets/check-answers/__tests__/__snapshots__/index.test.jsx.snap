// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#sprockets/sprockets/check-answers <Sprocket /> With additional props matches the snapshot 1`] = `
<div
  className="MOCK CLASSNAME"
>
  <CheckAnswersTitle>
    <h2
      className="MOCK CLASSNAME"
    >
      <TextContent>
        <span
          className="text-content"
        >
          MOCK TITLE
        </span>
      </TextContent>
    </h2>
  </CheckAnswersTitle>
  <CheckAnswersGroup>
    <dl
      className="MOCK CLASSNAME"
    >
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK STRING TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK STRING VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-string-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK STRING CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock string visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK NUMBER TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK NUMBER VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-number-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK NUMBER CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock number visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK OBJECT TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK OBJECT VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-object-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK OBJECT CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock object visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK ARRAY TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK ARRAY VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-array-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK ARRAY CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock array visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK BOOLEAN TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK BOOLEAN VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-boolean-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK BOOLEAN CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock boolean visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
      <div
        className="answer"
      >
        <AnswerTitle>
          <dt
            className="answer-title"
          >
            MOCK NULL TITLE
          </dt>
        </AnswerTitle>
        <AnswerValue>
          <dd
            className="answer-value"
          >
            MOCK NULL VALUE
          </dd>
        </AnswerValue>
        <ChangeAnswer>
          <dd
            className="change-answer"
          >
            <MockLink>
              <a
                className="mock-link"
                href="#mock-null-change-href"
              >
                <LinkText>
                  <TextContentWithVisuallyHiddenText>
                    <TextContent>
                      <span
                        className="text-content"
                      >
                        MOCK NULL CHANGE TEXT
                      </span>
                    </TextContent>
                     
                    <VisuallyHiddenText>
                      <span
                        className="visually-hidden-text"
                      >
                        mock null visually hidden text
                      </span>
                    </VisuallyHiddenText>
                  </TextContentWithVisuallyHiddenText>
                </LinkText>
              </a>
            </MockLink>
          </dd>
        </ChangeAnswer>
      </div>
    </dl>
  </CheckAnswersGroup>
</div>
`;

exports[`#sprockets/sprockets/check-answers <Sprocket /> With required props matches the snapshot 1`] = `null`;
